assets/cluster-monitoring-operator/prometheus-rule.yaml

```diff
@@ -382,6 +382,34 @@ spec:
       labels:
         workload_type: deploymentconfig
       record: namespace_workload_pod:kube_pod_owner:relabel
+    - alert: ThanosQueryOverload
+      annotations:
+        description: Thanos Query pod {{ $labels.pod }} in {{ $labels.namespace }}
+          namespace have been overloaded for more than 15 minutes. This may be a symptom
+          of excessive simultanous complex requests, low performance of the Prometheus
+          API, or failures within these components. Assess the health of the Thanos
+          query instances, the connnected Prometheus instances, look for potential
+          senders of these requests and then contact support.
+        summary: Thanos query reaches its maximum capacity serving concurrent requests
+      expr: thanos_query_concurrent_gate_queries_max - thanos_query_concurrent_gate_queries_in_flight
+        < 1
+      for: 15m
+      labels:
+        severity: warning
+    - alert: PrometheusQueryOverload
+      annotations:
+        description: Prometheus pod {{ $labels.pod }} in {{ $labels.namespace }} namespace
+          have been overloaded for more than 15 minutes. This may be a symptom of
+          excessive simultanous complex requests, low performance of the storage backend,
+          or failures within these components. Assess the health of the Prometheus
+          instances, the connnected Prometheus instances, look for potential senders
+          of these requests and then contact support.
+        summary: Prometheus reaches its maximum capacity serving concurrent requests
+      expr: prometheus_engine_queries_concurrent_max{job="prometheus-k8s"} - prometheus_engine_queries{job="prometheus-k8s"}
```

Comment:
check https://www.robustperception.io/alerting-on-gauges-in-prometheus-2-0/ to avoid alert resets in case of failed scrapes. I'm not sure exactly how the expression should look like though (maybe `max_over_time((prometheus_engine_queries_concurrent_max{job="prometheus-k8s"} - prometheus_engine_queries)[5m:]) < 1`)